spring-integration-kafka/src/main/java/org/springframework/integration/kafka/inbound/KafkaInboundEndpoint.java

```diff
@@ -64,16 +65,24 @@ default void doWithRetry(RetryTemplate template, RecoveryCallback<?> callback, O
 
 		try {
 			template.execute(context -> {
-				context.setAttribute(CONTEXT_RECORD, data);
-				context.setAttribute(CONTEXT_ACKNOWLEDGMENT, acknowledgment);
-				context.setAttribute(CONTEXT_CONSUMER, consumer);
+				if (context.getRetryCount() == 0) {
```

Comment:
I wonder if we can bring back a `ConsumerRecord` param type instead of that `Object data` since in the end we agree that we don't support retry for batch.

Reply:
Yes, that's good.